src/test/mod.rs

```diff
@@ -0,0 +1,239 @@
+                    TestTuple::new(l.left, r.left)
+                })
+            };
+            let join_op = Apply2::new(
+                move |left: &FiniteHashMap<TestTuple, i64>,
+                      right: &FiniteHashMap<TestTuple, i64>| join(left, right),
+            );
+            circuit
+                .add_binary_operator(join_op, &source0, &source1)
+                .unwrap()
+                .apply(ToString::to_string)
+                .inspect(move |s: &String| actual_data.borrow_mut().push(s.clone()));
+        })
+        .unwrap();
+
+        for _ in 0..3 {
+            root.step().unwrap()
+        }
+        let expected_output = vec!["{}", "{}", "{(2,1)=>1}"];
+        assert_eq!(&expected_output, actual_data_clone.borrow().deref());
+    }
+
+    // Test transitive closure
+    #[test]
+    fn transitive_closure() {
+        let actual_data = Rc::new(RefCell::new(Vec::<String>::new()));
+        let actual_data_clone = actual_data.clone();
+        let root = Root::build(|circuit| {
+            let mut z = ZSetHashMap::<TestTuple, i64>::new();
+            let mut t = 0;
+            let gen = Generator::new(move || {
+                let result = z.clone();
+                if t == 0 {
+                    z.increment(&TestTuple::new(1, 2), 1);
+                    z.increment(&TestTuple::new(2, 3), 1);
+                } else if t == 1 {
+                    z.increment(&TestTuple::new(3, 4), 1);
+                } else if t == 2 {
+                    z.increment(&TestTuple::new(2, 3), -1);
+                }
+                t = t + 1;
+                result
+            });
+            let source0 = circuit.add_source(gen);
+            let circuit_output = circuit
+                .iterate(|child| {
+                    let ns = NestedSource::new(
+                        source0,
+                        child,
+                        move |z: &mut ZSetHashMap<TestTuple, i64>| *z = z.clone(),
+                    );
+                    let i_output = child.add_source(ns).integrate();
+                    let (z1_output, z1_feedback) =
+                        child.add_feedback(Z1::new(ZSetHashMap::<TestTuple, i64>::new()));
+                    let add_output = i_output.plus(&z1_output);
+                    let join_func =
+                        |z0: &ZSetHashMap<TestTuple, i64>, z1: &ZSetHashMap<TestTuple, i64>| {
+                            z0.join(z1, &|l| l.right, &|r| r.left, |l, r| {
+                                TestTuple::new(l.left, r.right)
+                            })
+                        };
+                    let join_op = Apply2::new(
+                        move |left: &ZSetHashMap<TestTuple, i64>,
+                              right: &ZSetHashMap<TestTuple, i64>| {
+                            join_func(left, right)
+                        },
+                    );
+                    let join_output = child
```

Comment:
we don't have a nice syntax for inserting an Apply2?

Reply:
Not yet.  I haven't come up with a nice name for the method.  Any suggestions?

Reply:
it is doing a zip of the streams and then a map/apply.
ZipMap.

Reply:
`zip_with()` maybe?